### PR TITLE
This PR fixes how comments are queried and parsed.

### PR DESCRIPTION
**Pull request by satyen-blink**
See https://blinkhealth.atlassian.net/browse/INFRAENG-3572 for full context.

To rebuild modules dependencies: `go mod tidy`

### Broken behavior
 Atlantis plan runs and gets updated.  It is not guaranteed to be the
 last comment posted on the PR.  The gh action looks ONLY at the last
 comment and is an incorrect assumption. If the last comment inspected
 is not the expected comment, auto approval fails, causing a timeout
 failure.
### Fix
 Search all the comments in reverse order, look for the matching string
 for atlantis 'plan/apply' and return the comment to operation on.  If
     the expected comment is not found, we timeout.
### Changes
 - remove artificial retries.
 - do not search all comments and then return all comments only to inspect the last one - insteadd search for the expected comment and return it.
 - Tune overall timeout to ~ 3 minutes of max elapsed time (with random interval variations)
 - update github api version in `main.go`

# Testing timeout (manually)
Artificially `continue` and not take any actions and let it time out.  MaxElapsed was set to 3 minutes.

```
 cat ~/Documents/gh-actions.test
 go build && time ./atlantis-gh-action 22
PROCESSING PR blinkhealth/test1-tf-ctx/pull/22
Retrieving comments...
github-actions[bot]
blinkhealthgithub
github-actions[bot]
blinkhealthgithub
github-actions[bot]
blinkhealthgithub
github-actions[bot]
blinkhealthgithub

// snip //

github-actions[bot]
blinkhealthgithub
why this error: Error: Unexpected error - reached Timeout of ~ 20.0 minutes.
panic: Error: Unexpected error - reached Timeout of ~ 20.0 minutes.

goroutine 1 [running]:
main.waitPlan({0x14000020162, 0xb}, {0x1400002016e, 0xc}, 0x10474e1e0?)
	/Users/satyen.sheth/Documents/bh_repos/atlantis-gh-action/main.go:176 +0x344
main.main()
	/Users/satyen.sheth/Documents/bh_repos/atlantis-gh-action/main.go:233 +0x31c
./atlantis-gh-action 22  0.27s user 0.14s system 0% cpu 19:49.56 total
```

# Testing full run
Expected to fail because can't self-approve PR
```
go build && time ./atlantis-gh-action 22
PROCESSING PR blinkhealth/test1-tf-ctx/pull/22
Retrieving comments...
Skipping: user: github-actions[bot] comment created at:[2023-08-01 22:40:41 +0000 UTC] pr created at:[2023-08-01 22:39:47 +0000 UTC]
Result found: user: blinkhealthgithub comment created at:[2023-08-01 22:40:38 +0000 UTC] pr created at:[2023-08-01 22:39:47 +0000 UTC] time delta: 51
returning first line : Ran Plan for dir: `provisioning/live/dev/aws/eks-iam` workspace: `provisioning_live_dev_aws_eks-iam`
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -134,17 +177,21 @@ func waitPlan(org string, repo string, prNum int) string {
 			// for other errors, abort execution
 			panic(errors.New(errorStr))
 		}
+		fmt.Printf("ended up here: %v\n")
```
Comment (ganeshan-blink) [inline main.go:180]: Was this for debugging and do you need these comments ?

File: main.go [hunk 2]
```diff
@@ -10,14 +10,41 @@ import (
...
 
 	"github.com/cenkalti/backoff/v4"
-	"github.com/google/go-github/v39/github"
+	"github.com/google/go-github/v53/github"
 	"golang.org/x/oauth2"
 )
 
+const (
+	TimeOut             = 2 * time.Second
+	initialInterval     = 800 * time.Millisecond
```
Comment (hndrewaall) [inline main.go:19]: your caps styles are mixed. you should pick either `LikeThis` or `likeThis` and stick to it IMO

Comment (hndrewaall) [reply main.go:19]: from googling it seems that in golang initial caps => exported. so these should all be `likeThis`

File: main.go [hunk 3]
```diff
@@ -10,14 +10,41 @@ import (
...
+)
+
 var client *github.Client
 var ctx context.Context = context.Background()
 var atlantisPath string
 
+func getPrCreatedAt(ctx context.Context, client github.Client, org string, repo string, prNum int) github.Timestamp {
+	var pr *github.PullRequest
+	var err error
+	pr, _, err = client.PullRequests.Get(ctx, org, repo, prNum)
```
Comment (hndrewaall) [inline main.go:41]: i don't think you need the explicit type declarations, and even if you want them, i think you can make this all one line

```suggestion
	pr *github.PullRequest, _, err error := client.PullRequests.Get(ctx, org, repo, prNum)
```

File: main.go [hunk 4]
```diff
@@ -44,55 +71,75 @@ func approvePr(org string, repo string, prNum int) {
...
 
 func waitForComment(ctx context.Context, client github.Client, org string, repo string, prNum int, match string, errorMatch string, max_tries int) (*github.IssueComment, error) {
-	opt_cmt := &github.IssueListCommentsOptions{}
-
-	var comments []*github.IssueComment
-	var err error
-	i := 1
 
-	fmt.Printf("Retrieving comments .")
+	// EMPTY options - because it don't work!
```
Comment (hndrewaall) [inline main.go:75]: I don't understand this comment. Can you elaborate?

File: main.go [hunk 5]
```diff
@@ -44,55 +71,75 @@ func approvePr(org string, repo string, prNum int) {
...
+	opt := &github.IssueListCommentsOptions{}
 
-	f := func() error {
+	fmt.Println("Retrieving comments...")
 
-		comments, _, err = client.Issues.ListComments(ctx, org, repo, prNum, opt_cmt)
+	// callback to pass to the retry
+	f := func() (*github.IssueComment, error) {
```
Comment (hndrewaall) [inline main.go:81]: single char variables names make me sad :(

can you use something longer/more greppable for future maintainability? preferably with semantic value (like say `issueCommentCallback`)

File: main.go [hunk 6]
```diff
@@ -44,55 +71,75 @@ func approvePr(org string, repo string, prNum int) {
...
+		prc := getPrCreatedAt(ctx, client, org, repo, prNum)
```
Comment (hndrewaall) [inline main.go:82]: not single char, but still a bit obtuse. I don't know looking at this what kind of object this is. please make a bit more explicit

File: main.go [hunk 7]
```diff
@@ -44,55 +71,75 @@ func approvePr(org string, repo string, prNum int) {
...
-				return nil
+        /* Iterate over the comments in reverse order because github api sux
+         * or I can't seem to figure out how to sort by date. FWIW, 'curl' call
+         * returns the same order as this code does - and implies their API
+         * does not honor query parameters - giving up and explore comments
+         * backwards
+         */
+		for idx := len(comments) - 1; idx >= 0; idx-- {
+			c := comments[idx]
+			u := c.GetUser()
```
Comment (hndrewaall) [inline main.go:97]: no more single char vars!

**Review by hndrewaall — APPROVED** [review]
Looks good! I added a bunch of little style nits but I trust you to fix them without me double checking. But feel free to request a re-review if you want

**Review by satyen-blink — COMMENTED** [review]
all comments addressed.

**Pull request merged**